src/ILLink.RoslynAnalyzer/TrimAnalysis/FlowAnnotations.cs

```diff
@@ -23,31 +24,34 @@ sealed partial class FlowAnnotations
 
 		public static bool RequiresDataFlowAnalysis (IMethodSymbol method)
 		{
-			if (method.GetDynamicallyAccessedMemberTypes () != DynamicallyAccessedMemberTypes.None)
-				return true;
-
 			if (GetMethodReturnValueAnnotation (method) != DynamicallyAccessedMemberTypes.None)
 				return true;
 
-			foreach (var parameter in method.Parameters) {
-				if (GetMethodParameterAnnotation (parameter) != DynamicallyAccessedMemberTypes.None)
+			foreach (var param in method.GetParameters ()) {
+				if (GetMethodParameterAnnotation (param) != DynamicallyAccessedMemberTypes.None)
 					return true;
 			}
 
 			return false;
 		}
 
-		public static DynamicallyAccessedMemberTypes GetMethodParameterAnnotation (IParameterSymbol parameter)
+		public static DynamicallyAccessedMemberTypes GetMethodParameterAnnotation (ParameterProxy param)
 		{
+			IMethodSymbol method = param.Method.Method;
+			if (param.IsImplicitThis)
+				return method.GetDynamicallyAccessedMemberTypes ();
+
+			IParameterSymbol parameter = param.ParameterSymbol!;
 			var damt = parameter.GetDynamicallyAccessedMemberTypes ();
 
 			// Is this a property setter parameter?
-			var parameterMethod = (IMethodSymbol) parameter.ContainingSymbol;
-			Debug.Assert (parameterMethod != null);
+			Debug.Assert (method != null);
 			// If there are conflicts between the setter and the property annotation,
 			// the setter annotation wins. (But DAMT.None is ignored)
-			if (parameterMethod!.MethodKind == MethodKind.PropertySet && damt == DynamicallyAccessedMemberTypes.None) {
-				var property = (IPropertySymbol) parameterMethod.AssociatedSymbol!;
+			if (method!.MethodKind == MethodKind.PropertySet && damt == DynamicallyAccessedMemberTypes.None
+				// Indexer setters (.Item[int].set) will have the indexer as the first param and the value as the last. We only want annotations applied to the value parameter.
+				&& (int) param.Index + 1 == param.Method.GetParametersCount ()) {
```

Comment:
Since there could be multiple indices on the indexer, we'll need to check if `param` is the last parameter if it is an indexer. We could only enforce that check if the method is an indexer, but we'd have to cast to IPropertySymbol. Depending on how expensive a cast is, it might not gain us much.